cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala

```diff
@@ -339,16 +340,14 @@ class SnappyUnifiedMemoryManager private[memory](
   }
 
   private def getMinHeapEviction(required: Long): Long = {
-    // evict at least 100 entries to reduce GC cycles
     val waitingThreads = threadsWaitingForStorage.get()
-    math.max(required * waitingThreads, math.min(minHeapEviction,
-      required * math.max(100L, waitingThreads + 1)))
+    math.max(required, math.min(math.max(minHeapEviction, required * waitingThreads),
+      maxHeapEviction))
   }
 
   private def getMinOffHeapEviction(required: Long): Long = {
     // off-heap calculations are precise so evict exactly as much as required
-    (required * math.max(1, math.min(4, threadsWaitingForStorage.get()))) -
-        offHeapStorageMemoryPool.memoryFree
+    math.max(0, required - offHeapStorageMemoryPool.memoryFree + (1024 * 1024))
```

Comment:
What is the reason behind evicting 1 MB extra? Also at least 4 times the required - free was being evicted previously and not now? What am I missing?

Reply:
Earlier it was evicting too much especially because the putAll would pre-allocate bytes for all the values being put (and then thread would evict 4 times of even that).

1M is just arbitrary number to ensure that eviction does not fall short by just a small amount because putAll pre-allocation is an estimate.